### PR TITLE
drivers: modem: ublox-sara-r4: autodetection of modem and APN

### PR DESCRIPTION
**Pull request by hwilmers**
Implemented:
- automatic detection of modem type (Sara R410 or Sara U201)
- automatic setting of APN, depending on inserted SIM card

Automatic detection of the modem type:

After hardware initialisation of the modem, the type of modem
is detected. After this, further initialisation and functionality
is chosen depending on the detected modem type.

To enable this functionality, set following bool in Kconfig:
   MODEM_UBLOX_SARA_AUTODETECT_VARIANT

Automatic setting of APN:

During communication initialisation, the IMSI of the inserted SIM
card is evaluated to determine the APN. This is done by comparing
the first 5 characters of the IMSI to a list of known providers.
The list can be given in Kconfig.

To enable this functionality, set following bool in Kconfig:
   MODEM_UBLOX_SARA_AUTODETECT_APN

To set a list of providers, set following string:
   MODEM_UBLOX_SARA_AUTODETECT_APN_PROFILES

If the provider can not be found in the list, the APN given in
following entry is used as a fallback:
   MODEM_UBLOX_SARA_R4_APN

Signed-off-by: Hans Wilmers <hans@wilmers.no>

### CODE REVIEW

**Comment by hwilmers** [comment]
Caveat: When autodetection is chosen, it is presumed that initialisation of the hardware can be identical for R410 and U201. On my hardware this is the case, but probably it is not true generally.

File: drivers/modem/ublox-sara-r4.c
Comment (jukkar) [inline drivers/modem/ublox-sara-r4.c]: Just squash the "drivers: modem: ublox-sara-r4: removed blank line" commit to previous one.

**Review by jukkar — COMMENTED** [review]
We have also the generic GSM modem support and something like this could be implemented there too (if using ublox-sara modem in non-offloading case). In order not to duplicate too much code, I just wonder if we could have some generic way of adding this kind of support.

File: drivers/modem/ublox-sara-r4.c
Comment (hwilmers) [reply drivers/modem/ublox-sara-r4.c]: @jukkar Sorry - tried to do that, but it only added more commits. Should I clear this up some way or another?

Comment (jukkar) [reply drivers/modem/ublox-sara-r4.c]: Just squash the commits together, and force push to this branch. Do not close this PR and resubmit, it is just pointless.

Comment (jukkar) [reply drivers/modem/ublox-sara-r4.c]: Are you trying to do this from github web ui? That (squashing, force pushing) is probably not going to work from web ui, one needs to use git command line for that.

Comment (hwilmers) [reply drivers/modem/ublox-sara-r4.c]: got it - thanks for the comments.

File: drivers/modem/Kconfig.ublox-sara-r4
Comment (jukkar) [inline drivers/modem/Kconfig.ublox-sara-r4]: Please indent the help lines like the other help blocks, so with two spaces

File: drivers/modem/ublox-sara-r4.c
Comment (jukkar) [inline drivers/modem/ublox-sara-r4.c]: perhaps static here?

File: drivers/modem/ublox-sara-r4.c
Comment (jukkar) [inline drivers/modem/ublox-sara-r4.c]: Introduce the variable at the beginning of the block, not middle.

File: drivers/modem/ublox-sara-r4.c
Comment (jukkar) [inline drivers/modem/ublox-sara-r4.c]: If we have an error, should we bail out here?

File: drivers/modem/ublox-sara-r4.c
Comment (jukkar) [inline drivers/modem/ublox-sara-r4.c]: Instead of this hard coded value, I think something like below would be better (user does not have to count the chars itself as the compiler is doing that for you). The APN name just needs to be the longest name possible.
```
char cmd[sizeof("AT+CGDCONT=1,'IP','012345678901234567890123456789012'")];
```

File: drivers/modem/ublox-sara-r4.c
Comment (jukkar) [inline drivers/modem/ublox-sara-r4.c]: Similar thing here

Comment (hwilmers) [reply drivers/modem/ublox-sara-r4.c]: Thank you for your input. I will look at this within a few days. It it ok for the workflow if I add commits, or should I squash them again, replacing the original commit?

Comment (jukkar) [reply drivers/modem/ublox-sara-r4.c]: As a general rule, if you are fixing things in your own commits, then squash. If you are introducing new logically separate stuff, then do new commit. In this case, just squash the commits.

File: drivers/modem/ublox-sara-r4.c
Comment (hwilmers) [reply drivers/modem/ublox-sara-r4.c]: I need to be able to change this on application startup. The least intrusive way I see is to use global scope - but if we should do it in another way, please comment.

File: drivers/modem/ublox-sara-r4.c
Comment (hwilmers) [reply drivers/modem/ublox-sara-r4.c]: we don't bail out here, because we want to restart a new RSSI query, regardless of teh outcome of this one. This is also how it is done in the original code.

File: drivers/modem/Kconfig.ublox-sara-r4
Comment (hwilmers) [reply drivers/modem/Kconfig.ublox-sara-r4]: done

File: drivers/modem/ublox-sara-r4.c
Comment (hwilmers) [reply drivers/modem/ublox-sara-r4.c]: done

File: drivers/modem/ublox-sara-r4.c
Comment (hwilmers) [reply drivers/modem/ublox-sara-r4.c]: not really - we want to start the next query afterwards, regardless the outcome of this one.

File: drivers/modem/ublox-sara-r4.c
Comment (hwilmers) [reply drivers/modem/ublox-sara-r4.c]: good point! I changed it to use the correct (max) length now.

**Comment by jukkar** [comment]
ping @mike-scott

**Comment by jukkar** [comment]
@mike-scott does this look ok to you?

**Comment by mike-scott** [comment]
@jukkar I'll play test today.  Thanks for the reminder.

**Comment by mike-scott** [comment]
@hwilmers After reviewing this, I think this should be 2 commits:
1. automatic detection of modem type (Sara R410 or Sara U201) changes
2. and then automatic setting of APN, depending on inserted SIM card

I also am questioning the complexity of the #ifdefs in this driver. If we can come up with a creative solution for the different pin init between U2 and R4, then we could do away with the manual selection of the modem type entirely.

But let's get the 2 changes into their own commits first.

**Comment by hwilmers** [comment]
@mike-scott Changes are now divided in two commits. Would you have any suggestions about how to reduce complexity of the #ifdefs? I am reluctant to introduce dead code, but if it would be ok to introduce some, complexity could be reduced.

File: drivers/modem/ublox-sara-r4.c
Comment (jukkar) [inline drivers/modem/ublox-sara-r4.c:834]: Shouldn't these two vars be const instead of static?

File: drivers/modem/ublox-sara-r4.c
Comment (jukkar) [inline drivers/modem/ublox-sara-r4.c:839]: also this could be const

File: drivers/modem/ublox-sara-r4.c
Comment (jukkar) [inline drivers/modem/ublox-sara-r4.c:847]: indentation

File: drivers/modem/ublox-sara-r4.c
Comment (jukkar) [inline drivers/modem/ublox-sara-r4.c:279]: Indentation. Is your tab size correct, it should be 8

File: drivers/modem/ublox-sara-r4.c
Comment (jukkar) [inline drivers/modem/ublox-sara-r4.c:284]: ditto

File: drivers/modem/ublox-sara-r4.c
Comment (jukkar) [inline drivers/modem/ublox-sara-r4.c:1010]: ditto

File: drivers/modem/ublox-sara-r4.c
Comment (jukkar) [inline drivers/modem/ublox-sara-r4.c:1016]: here too

File: drivers/modem/ublox-sara-r4.c
Comment (jukkar) [inline drivers/modem/ublox-sara-r4.c:1107]: and here

**Comment by mike-scott** [comment]
@hwilmers Thanks for splitting this up.  I was thinking about the exact same thing with the `#ifdef` complexity.  If we could consolidate the init process for both R4 and U2, then we would remove the 3 `#ifdef`s which are really the only sticking point to removing the variant `#ifdef`s entirely.  It would increase the code size of the driver by an unknown amount (not a huge amount) -- but would also provide good code coverage.

**Comment by mike-scott** [comment]
@hwilmers Let me look at the Application notes power on/off sequence for both the R4xx and U2xx and see if we can conslidate.

**Comment by hwilmers** [comment]
@mike-scott Is there a chance to get this in before the merge window closes?

**Comment by jukkar** [comment]
I think we could try to get this to 2.3, @mike-scott please ack if ok.

**Comment by carlescufi** [comment]
@jukkar I'd need at least one approval for this to go in. I will add @rlubos as well in case he can take a look.

**Comment by mike-scott** [comment]
I'll take another look

**Comment by mike-scott** [comment]
@hwilmers @jukkar @carlescufi I think this PR still needs a bit of discussion:

My current issues are:
1. The R4 vs. U2 differences in `pin_init()` are not addressed.  I think in it's current form, this PR would probably break R4.
2. The second patch might be mergeable as a stand-alone patch, but it only applies to U2 which uses the AT+UPSD command.  R4 relies on AT+CGDCONT to set APN.

**Comment by hwilmers** [comment]
>     1. The R4 vs. U2 differences in `pin_init()` are not addressed.  I think in it's current form, this PR would probably break R4.
> 
Yes, I see that there is a problem. But it is the other way around, since with this patch (but only with MODEM_UBLOX_SARA_AUTODETECT_VARIANT set), for both modems the pin_init() is done as though they were R4. I have no better solution for this, and because on my hardware pin_init() is merely ignored, I can not test for this case.

>     2. The second patch might be mergeable as a stand-alone patch, but it only applies to U2 which uses the AT+UPSD command.  R4 relies on AT+CGDCONT to set APN.

If R4 is detected, AT+CGDCONT is used to set the APN. I apologise for the complicated #ifdef logic that makes this less obvious than it could be.

**Comment by mike-scott** [comment]
@hwilmers 
> > ```
> > 1. The R4 vs. U2 differences in `pin_init()` are not addressed.  I think in it's current form, this PR would probably break R4.
> > ```
> 
> Yes, I see that there is a problem. But it is the other way around, since with this patch (but only with MODEM_UBLOX_SARA_AUTODETECT_VARIANT set), for both modems the pin_init() is done as though they were R4. I have no better solution for this, and because on my hardware pin_init() is merely ignored, I can not test for this case.

I would also hate to break U2 users.  I don't think we can merge the first patch until we figure out the pin_init changes. :/

> 
> > ```
> > 2. The second patch might be mergeable as a stand-alone patch, but it only applies to U2 which uses the AT+UPSD command.  R4 relies on AT+CGDCONT to set APN.
> > ```
> 
> If R4 is detected, AT+CGDCONT is used to set the APN. I apologise for the complicated #ifdef logic that makes this less obvious than it could be.

Oh you are correct.  The mdm_apn value is also used for AT+CGDCONT (somehow I missed this initially).  So the 2nd patch can probably be merged.  Do you want to submit that as a separate patch in a new PR or perhaps remove the first patch from here and re-post that patch as a new PR so it can continue to develop?

**Comment by mike-scott** [comment]
@hwilmers The first patch is going to need the following:
1. A U2 user to confirm whatever pin_init changes are working for them
2. I can validate the R4 pin_init changes on my end

**Comment by hwilmers** [comment]
@mike-scott About the first patch: I may be missing something here, but there are no changes in pin_init() at all. If you choose to set MODEM_UBLOX_SARA_U2 or MODEM_UBLOX_SARA_R4, it will behave exactly as before.

But I don't want to force this onto anyone. If you think it is better, I can remove this patch from the PR.

**Comment by mike-scott** [comment]
> @mike-scott About the first patch: I may be missing something here, but there are no changes in pin_init() at all. If you choose to set MODEM_UBLOX_SARA_U2 or MODEM_UBLOX_SARA_R4, pin_init() will behave exactly as before.
> 
> But I don't want to force this onto anyone. If you think it is better, I can remove this patch from the PR.

@hwilmers I should have worded my original response this way:

I wouldn't want to merge the first patch and have users try it with unexpected results.  I would rather merge it when it's ready for all users.

You are correct in that by itself it doesn't actually break anyone.  As long as they don't swap to using `MODEM_UBLOX_SARA_AUTODETECT_VARIANT`.

I'm curious how you are able to use the first patch as it is now.  You mention that for your hardware pin_init() is ignored?  What HW is that?  If there is a valid user of this patch, then I would be more inclined to merge it now and then work to cleanup `pin_init()`.

**Comment by hwilmers** [comment]
> I'm curious how you are able to use the first patch as it is now. You mention that for your hardware pin_init() is ignored? What HW is that? If there is a valid user of this patch, then I would be more inclined to merge it now and then work to cleanup `pin_init()`.
I am using cellular XBee modules made by Digi. The hardware init of the modem engines is done by by some middle layer made by Digi, but after hardware init they allow to communicate directly to the modem engines (in this case U201 or R410). This is why pin_init() works identical for both types - it is merely ignored.

**Comment by mike-scott** [comment]
@hwilmers if you say that this PR makes the cellular XBee modules made by Digi usable with Zephyr.  Then I'm ok with +1 for merge in 2.3.

**Review by mike-scott — APPROVED** [review]
Let's add this for 2.3 to enable support for cellular XBee modules made by Digi.

**Pull request merged**
